doc/developer/design/20230802_explain_running_dataflows.md

```diff
@@ -0,0 +1,255 @@
+
+# Motivation
+[motivation]: #motivation
+
+Various troubleshooting and optimization techniques for long-running dataflows depend on looking and interpreting the optimized and physical plans for the index or materialized view associated with that dataflow.
+At the moment, both customers and field engineering rely on
+
+```sql
+EXPLAIN [OPTIMIZED|PHYSICAL] PLAN FOR MATERIALIZED VIEW
+```
+
+to get this information for materialized views, and we still don't support the counterpart syntax for index items.
+
+However, even for the `MATERIALIZED VIEW` syntax, the information can be misleading because the `EXPLAIN` output is based on an a new optimization run that is executed while handling the `EXPLAIN` statement.
+Even though our optimizer is deterministic, this will only provide results consistent with the dataflow that is actually running while the environment context that guides the optimizer remains unchanged.
+For example, if the user does:
+
+```sql
+-- The definition of V optimized with an empty set of indexes.
+CREATE MATERIALIZED VIEW V AS (SELECT * FROM T where a = 5);
+-- A new index on T(a) is created.
+CREATE T_a INDEX ON T(a);
+-- The explain is produced from an optimization run that assumes that an index on T(a) exists.
+EXPLAIN MATERIALIZED VIEW V;
+```
+
+The `EXPLAIN` output will indicate that we are using `T_a` for the point lookup, while in practice this will not be the case because the index was created _after_ installing the dataflow that maintains `V`.
+
+This design doc proposes a sequence of changes to the catalog and our `EXPLAIN` handling code that would allow us to change the behavior of
+
+```sql
+EXPLAIN [OPTIMIZED|PHYSICAL] PLAN FOR MATERIALIZED VIEW
+```
+
+and implement similar behavior for
+
+```sql
+EXPLAIN [OPTIMIZED|PHYSICAL] PLAN FOR INDEX
+```
+
+# Explanation
+[explanation]: #explanation
+
+At the moment, `sequence_explain_plan` constructs its output using information derived from optimizations done at the time when `sequence_explain_plan` runs.
+
+At the high level, the API that constructs an `EXPLAIN` output for a given plan looks like that:
+
+```rust
+let context = ExplainContext {
+    config,
+    humanizer,
+    used_indexes,
+    finishing,
+    duration,
+};
+
+Explainable::new(&mut plan).explain(format, context)
+```
+
+We need to
+(1) analyze which values in the above code snippet represent information that should reflect the state of the world when the dataflow was originally created, and
+(2) suggest mechanisms to make these values available for subsequent `EXPLAIN <catalog item>` calls.
+
+In general, we can persist the required data
+(a) in-memory, as extensions of the `CatalogItem` objects that are maintained as entries in the `CatalogState`, or
+(b) durably, as extensions of the `SerializedCatalogItem` structures that represent the part of the catalog state that survives restarts.
```

Comment:
FWIW, this is the full struct that is stored on disk for catalog items: https://github.com/MaterializeInc/materialize/blob/a0be13cb9dd52c6e60faa500b74d631bdbdcc515/src/adapter/src/catalog/storage.rs#L2049-L2061

Reply:
Thanks. I've also referenced this in "Appendix B".